packages/drivers/odsp-driver/src/fetchSnapshot.ts

```diff
@@ -205,7 +207,7 @@ async function fetchLatestSnapshotCore(
             },
             async (event) => {
                 const startTime = performance.now();
-                const response: IOdspResponse<IOdspSnapshot> = await snapshotDownloader(
+                const response = await snapshotDownloader<IOdspSnapshot>(
```

Comment:
Would it be better to push requirement to convert into snapshotDownloader, such that this function and others are abstracted away from wire format? That way plugging in new format would be much simpler (as of now, all of these functionality would need to be duplicated).
Basically there are three layers:
1. Generic fetch functionality - does not know what it fetches
2. Middle layer, that works with specific wire format and does conversion to "common" format. This layer is ideally just one function.
3. All the rest of the code (majority of it) that does not know anything about wire format.

This function I believe belongs to # 3.

Reply:
Yes in this: snapshotDownloader belongs to #1. fetchSnapshotCore belongs to #2 and all code in storage manager is 3.

Reply:
"snapshotDownloader" is just the stub which fetches the snapshot(initiates network call). It does not know what type of snapshot it is and what conversion needs to be done. If we want snapshotDownloader to do the conversion, then that conversion logic will be passed by "fetchLatestSnapshotCore" to snapshotDownloader as callback or maybe some arrgument telling what type of snapshot it is so that it can do the conversion as only "fetchLatestSnapshotCore " knows which type of snapshot is fetched.

Reply:
So if I got it right, we would need to rename fetchLatestSnapshotCore into something like fetchLatestSnapshotV1, add parallel fetchLatestSnapshotV2, and make fetchSnapshotWithRedeem() be the function that calls both versions depending on feature flag, right?

fetchLatestSnapshotCore is called twice from fetchSnapshotWithRedeem and fetchLatestSnapshotCore  has a lot of logic (resource timing telemetry) that would be useful for both paths, so I'd probably try to create fetchLatestSnapshotV1 & fetchLatestSnapshotV2 be responsible only for URI formatting and initial parsing / conversion of data, but keep fetchLatestSnapshotCore as a function on top of them (and the one where we do fork between two calls).

Reply:
fetchSnapshotWithRedeem would just call to fetchLatestSnapshotCore. It is fetchLatestSnapshotCore which will decide what to fetch. I can add fetchLatestSnapshotV1 and V2 once I integrate the second format.